### PR TITLE
ONNXConstantOp custom assembly format

### PR DESCRIPTION
**Pull request by sorenlassen**
This makes the mlir assembly more concise and easier to read.

It also opens the door to print alternatives to DenseElementsAttr, like DenseResourceElementsAttr, in the same way as DenseElementsAttr to hide these internal representations.

### CODE REVIEW

**Review by tungld — COMMENTED** [review]
I like the new format. Thanks!

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -127,3 +127,93 @@ Type getBroadcastedRankedType(
...
+  Type resultType = getResult().getType();
+  if (auto attr = value()) {
+    // ONNXConstantOp value must be ElementsAttr, but not SparseElementsAttr.
+    auto elements = attr->cast<ElementsAttr>();
+    assert(!elements.isa<SparseElementsAttr>() &&
+           "ONNXConstantOp value cannot be sparse");
+    if (elements.getType() == resultType) {
+      // NOTE: Here we can insert logic to print alternatives to
+      //       DenseElementsAttr, like DenseResourceElementsAttr, in the same
+      //       way as DenseElementsAttr to hide these internal representations.
```
[the remark this answers is not in the document]
Comment (sorenlassen) [reply src/Dialect/ONNX/ONNXOps.cpp:172]: yes

File: test/mlir/onnx/onnx_decompose_einsum.mlir [hunk 2]
```diff
@@ -107,12 +107,12 @@ func.func @test_einsum_ibh_hnd(%arg0: tensor<128x1x1024xf16>, %arg1: tensor<1024
   return %0 : tensor<128x1x16x64xf16>
 // CHECK-LABEL:  func.func @test_einsum_ibh_hnd
 // CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<128x1x1024xf16>, [[PARAM_1_:%.+]]: tensor<1024x16x64xf16>) -> tensor<128x1x16x64xf16> {
-// CHECK:           [[VAR_0_:%.+]] = "onnx.Constant"() {value = dense<[128, 1024]> : tensor<2xi64>} : () -> tensor<2xi64>
+// CHECK:           [[VAR_0_:%.+]] = onnx.Constant dense<[128, 1024]> : tensor<2xi64>
 // CHECK-DAG:       [[VAR_1_:%.+]] = "onnx.Reshape"([[PARAM_0_]], [[VAR_0_]]) {allowzero = 0 : si64} : (tensor<128x1x1024xf16>, tensor<2xi64>) -> tensor<128x1024xf16>
-// CHECK-DAG:       [[VAR_2_:%.+]] = "onnx.Constant"() {value = dense<1024> : tensor<2xi64>} : () -> tensor<2xi64>
+// CHECK-DAG:       [[VAR_2_:%.+]] = onnx.Constant dense<1024> : tensor<2xi64>
```
Comment (sorenlassen) [inline test/mlir/onnx/onnx_decompose_einsum.mlir:112]: @tungld here is an example of a splat constant

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 3]
```diff
@@ -127,3 +127,93 @@ Type getBroadcastedRankedType(
...
+//
+// In case of type mismatch or if the op has another attr than sparse_value or
+// value or has no attr, the op prints an attribute dictionary followed by the
+// op result type, just like the default assembly format.
+//
+//   onnx.Constant {value = dense<1024> : tensor<1xi64>} : tensor<*xi64>
+//
+//   onnx.Constant {value_int = 1 : si64} : tensor<i64>
+//
+//   onnx.Constant : tensor<64xf32>
```
[the remark this answers is not in the document]
Comment (sorenlassen) [reply src/Dialect/ONNX/ONNXOps.cpp:156]: splat DenseElementsAttr is written with its full tensor type `dense<1> : tensor<5x7x9xi64>` so the default print was
```
"onnx.Constant" {value = dense<1> : tensor<5x7x9xi64>} : tensor<5x7x9xi64>
```
and in the custom assembly format becomes
```
onnx.Constant dense<1> : tensor<5x7x9xi64>
```

File: test/mlir/onnx/onnx_decompose_einsum.mlir (hunk 2 above)
Comment (tungld) [reply test/mlir/onnx/onnx_decompose_einsum.mlir:112]: I see. Thanks!

**Review by tungld — APPROVED** [review]
LGTM!

**Comment by sorenlassen** [comment]
@chentong319 @AlexandreEichenberger heads up that I plan to land this PR later today - it changes all the constants in the lit tests so any other existing PRs that edit lit tests are likely to encounter merge conflicts

**Pull request merged**